### ISSUE TITLE
Resources still retrieved when expressly rejected using callbacks when using authenticated proxies

### ISSUE THREAD

**Issue by wflanagan** [rubycdp/ferrum#149]
Loving the gem so far. Much easier to deal with Ruby.

I ran into this problem when I was attempting to troubleshoot other things. 

I'm using callbacks to limit what is retrieved to limit bandwidth consumption (images, etc). I'm also using proxies.

In testing, I realized that my screenshots still had images that are expressly rejected in the callback. I added some code to show that the DENY was still firing in the callback.  

But, the images were still there. 

So, I wrote a separate script that can be used to test this case (below). 

If you don't use a proxy, everything works as expected and images are not present in the screenshot. 

If you add an authenticating proxy, the images and other callbacks show that they fire and are rejected. But, the images are still present. 

It's possible that I'm misconfiguring this. If so, please help me sort out what I'm doing wrong. But, if not, then it's a bug. 

In the code below, I show this. You can comment out the proxy details and re-run the script. Without a proxy, it will reject the images. If you put a proxy in, the images will still be present in the screenshot. 

If I've messed up my config and it's a misconfiguration problem, please let me know! 

```
require 'bundler/inline'

gemfile do
  source 'https://rubygems.org'
  gem 'ferrum'
end

require 'ferrum'

allowed_types = %w[Document]
headers = {"User-Agent"=>"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.88 Safari/537.36"}

# Remove the proxy server option from the hash to test without the proxy server
retriever_options = {:browser_options=>{"proxy-server"=>"XX:XX:XX:XX:39393", "ignore-certificate-errors"=>nil}, :timeout=>30, :xvfb=>false}

browser = ::Ferrum::Browser.new(retriever_options)
browser.headers.set(headers)

# Remove this line when you are testing without the proxy.
browser.network.authorize(type: :proxy, user: 'XXXXXXX', password: 'XXXXXXX')

browser.network.intercept
browser.on(:request) do |request|
  if allowed_types.include?(request.resource_type)
    puts "ALLOW #{request.resource_type} #{request.url}"
    request.continue
  else
    puts "DENY #{request.resource_type} #{request.url}"
    request.abort
  end
end

# Change this to the site you want to test with.
url = "https://site.com/"

browser.goto(url)
sleep 4.0 # I put this in just to make sure all JS is done rendering, doesn't affect the case. 
browser.screenshot(path: 'site.html.jpg')
```

**Comment by Mifrill** [comment]
Hello, @wflanagan !
Just try to move `browser.network.authorize(type: :proxy, user: 'XXXXXXX', password: 'XXXXXXX')` after the block `browser.on(:request)` declaration, like this:
```Ruby
browser.on(:request) do |request|
  if allowed_types.include?(request.resource_type)
    puts "ALLOW #{request.resource_type} #{request.url}"
    request.continue
  else
    puts "DENY #{request.resource_type} #{request.url}"
    request.abort
  end
end
browser.network.authorize(user: 'XXXXXXX', password: 'XXXXXXX')
```

Long story in details:
as I see, the point here in the `authorize` method itself, not in the proxy settings.

Because we already have `Subscriber` on `request` event inside `authorize` method:
https://github.com/rubycdp/ferrum/blob/7c84cdb14b1000fe6280614980187e605ed7718e/lib/ferrum/network.rb#L96-L98
As we can see here:
https://github.com/rubycdp/ferrum/blob/7c84cdb14b1000fe6280614980187e605ed7718e/lib/ferrum/browser/subscriber.rb#L19-L22
The `on` handler is adds blocks in a row (`@on[event] << block`)
so, that's why, in your case, we have 2 Subscribers with two handlers:
first handler as `request.continue` (inside `authorize` method) 
and second handler with your logic:
```Ruby
  if allowed_types.include?(request.resource_type)
    puts "ALLOW #{request.resource_type} #{request.url}"
    request.continue
  else
    puts "DENY #{request.resource_type} #{request.url}"
    request.abort
  end
```
btw, due to `authorize` method usage, we can avoid using `browser.network.intercept`, because `authorize` method already contains `intercept` call:
https://github.com/rubycdp/ferrum/blob/7c84cdb14b1000fe6280614980187e605ed7718e/lib/ferrum/network.rb#L94

**Comment by Mifrill** [comment]
@wflanagan please try this version:
```
gemfile do
  source 'https://rubygems.org'
  gem 'ferrum', git: "git@github.com:rubycdp/ferrum.git", branch: "bugfix/issue#149"
end
```

**Comment by Mifrill** [comment]
@route on thinking, it seems to me that the handler should be redefined, this is more explicit and predictable behavior, WDYT?

**Comment by wflanagan** [comment]
> @wflanagan please try this version:

I'll add it and let you know. Thanks for the feedback!

**Comment by wflanagan** [comment]
@Mifrill Thanks for working on this.  This does seem to work better. 

What do you think the timeline is to this type of thing getting merged into master? I am waiting on the fix for the warnings for `instance variable @xvfb not initialized` as well and it would be nice to get some of the mess out of my logs!

**Comment by Mifrill** [comment]
@wflanagan 

> This does seem to work better.

Good! So this issue is not related to proxies, it is the repeat-of-handlers issue.

> timeline is to this type of thing getting merged into master

well, technically, it's not needed, because, to resolve your specific case, you can use the current gem version with the change the order of `browser.on(:request)` and `browser.network.authorize`, as I said before.
so, in your code-snipped replace:
```
browser.network.authorize(type: :proxy, user: 'XXXXXXX', password: 'XXXXXXX')
...
browser.on(:request) do |request|
  if allowed_types.include?(request.resource_type)
...
```
into:
```
browser.on(:request) do |request|
  if allowed_types.include?(request.resource_type)
...
end

browser.network.authorize(type: :proxy, user: 'XXXXXXX', password: 'XXXXXXX')
```
It should work as you expect.

About:
> instance variable `@xvfb` not initialized

this issue is resolved #146 and merged into master, so you may use this way til the next release:
```
gemfile do
  source 'https://rubygems.org'
  gem 'ferrum', git: "git@github.com:rubycdp/ferrum.git", branch: "master"
end
```

**Comment by route** [comment]
@wflanagan fixed in master, see details in https://github.com/rubycdp/ferrum#authorizeoptions-block
